pkg/operator/controller/dnsnameresolver/resolver.go

```diff
@@ -0,0 +1,340 @@
+package dnsnameresolver
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+	networkv1alpha1 "github.com/openshift/api/network/v1alpha1"
+	"github.com/sirupsen/logrus"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	defaultMaxTTL    = 30 * time.Minute
+	defaultMinTTL    = 5 * time.Second
+	defaultPort      = "53"
+	maxCoreDNSPodIPs = 5
+)
+
+// dnsResolvedName contains details about a DNS resolved name.
+type dnsResolvedName struct {
+	minNextLookupTime time.Time
+	regularObjExists  bool
+	wildcardObjExists bool
+	numIPs            int
+}
+
+// Resolver keeps track of the DNS names, the corresponding IP addresses and TTLs. It sends DNS resolution
+// requests to CoreDNS pods whenever TTL of any IP address associated to a DNS name expires.
+type Resolver struct {
+	corednsEndpointSliceCache cache.Cache
+	serviceName               string
+	dnsNames                  map[string]dnsResolvedName
+	regularObjInfo            map[string]string
+	wildcardObjInfo           map[string]sets.Set[string]
+	dnsLock                   sync.Mutex
+
+	added   chan struct{}
+	deleted chan string
+}
+
+// NewResolver initializes and returns a new resolver instance.
+func NewResolver(cache cache.Cache, serviceName string) *Resolver {
+	return &Resolver{
+		corednsEndpointSliceCache: cache,
+		serviceName:               serviceName,
+		dnsNames:                  make(map[string]dnsResolvedName),
+		regularObjInfo:            make(map[string]string),
+		wildcardObjInfo:           make(map[string]sets.Set[string]),
+		added:                     make(chan struct{}, 1),
+		deleted:                   make(chan string, 1),
+	}
+}
+
+// Start starts a goroutine for sending DNS lookup requests of the DNS names on expiration of the corresponding TTLs.
+func (res *Resolver) Start() {
+	var nextDNSName, deletedDNSName string
+	var nextLookupTime time.Time
+	var exists bool
+	var numIPs int
+	// Initially the check for nextDNSName to be looked up will happen after default maximum TTL. If a DNS name
+	// gets added before that then the nextDNSName and timeTillNextLookup will be updated.
+	timeTillNextLookup := defaultMaxTTL
+	go func() {
+		timer := time.NewTicker(timeTillNextLookup)
+		defer timer.Stop()
+		for {
+			select {
+			case <-res.added:
```

Comment:
Why is it done this way instead of keeping watch on the dnsNameResolverCache?

Reply:
This is triggered by the Add call during the reconcile of a DNSNameResolver object. This indeed happens due to the watch on the dnsNameResolverCache.

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2